mesh/mesh.cpp

```diff
@@ -5774,6 +6154,754 @@ void Mesh::HexUniformRefinement()
+                                           oface+qf4, oelem+he, oface+qf3,
+                                           oedge+e[11], attr);
+            elements[j+3] = new Hexahedron(oedge+e[8], oface+qf1, oelem+he,
+                                           oface+qf4, v[4], oedge+e[4], oface+qf5,
+                                           oedge+e[7], attr);
+            elements[j+4] = new Hexahedron(oface+qf1, oedge+e[9], oface+qf2,
+                                           oelem+he, oedge+e[4], v[5], oedge+e[5],
+                                           oface+qf5, attr);
+            elements[j+5] = new Hexahedron(oelem+he, oface+qf2, oedge+e[10],
+                                           oface+qf3, oface+qf5, oedge+e[5], v[6],
+                                           oedge+e[6], attr);
+            elements[j+6] = new Hexahedron(oface+qf4, oelem+he, oface+qf3,
+                                           oedge+e[11], oedge+e[7], oface+qf5,
+                                           oedge+e[6], v[7], attr);
+
+            v[1] = oedge+e[0];
+            v[2] = oface+qf0;
+            v[3] = oedge+e[3];
+            v[4] = oedge+e[8];
+            v[5] = oface+qf1;
+            v[6] = oelem+he;
+            v[7] = oface+qf4;
+         }
+         break;
+         default:
+            MFEM_ABORT("MixedUniformRefinement() "
+                       "Unknown 3D element type \"" << elements[i]->GetType()
+                       << "\"");
+            break;
+      }
+   }
+
+   boundary.SetSize(4 * NumOfBdrElements);
+   for (i = 0; i < NumOfBdrElements; i++)
+   {
+      attr = boundary[i]->GetAttribute();
+      v = boundary[i]->GetVertices();
+      e = bel_to_edge->GetRow(i);
+      f = &be_to_face[i];
+      j = NumOfBdrElements + 3 * i;
+
+      if (boundary[i]->GetType() == Element::TRIANGLE)
+      {
+         boundary[j+0] = new Triangle(oedge+e[1], oedge+e[2], oedge+e[0], attr);
+         boundary[j+1] = new Triangle(oedge+e[0], v[1], oedge+e[1], attr);
+         boundary[j+2] = new Triangle(oedge+e[2], oedge+e[1], v[2], attr);
+
+         v[1] = oedge+e[0];
+         v[2] = oedge+e[2];
+      }
+      else if (boundary[i]->GetType() == Element::QUADRILATERAL)
+      {
+         int qf = (*f2qf_ptr)[f[0]];
+         boundary[j+0] = new Quadrilateral(oedge+e[0], v[1], oedge+e[1],
+                                           oface+qf, attr);
+         boundary[j+1] = new Quadrilateral(oface+qf, oedge+e[1], v[2],
+                                           oedge+e[2], attr);
+         boundary[j+2] = new Quadrilateral(oedge+e[3], oface+qf,
+                                           oedge+e[2], v[3], attr);
+
+         v[1] = oedge+e[0];
+         v[2] = oface+qf;
+         v[3] = oedge+e[3];
+      }
+      else
+      {
+         MFEM_ABORT("boundary Element is not a triangle or a quad!");
+      }
+   }
+
+   static const double A = 0.0, B = 0.5, C = 1.0;
+   static double tet_children[3*4*8] =
```

Comment:
Is this tetrahedron octasection? I guess there is no other choice in a mixed mesh...

Reply:
Yes, I believe it is.  There is the potential for generating bad tets (@v-dobrev brought this up recently) by repeated refinement unless we split the central octahedron in a more intelligent manner.

I wasn't sure of the best strategy to handle this so I chose the simplest option (always refine the same way).  We could potentially define different point matrices for the different tetrahedron refinement schemes or perhaps we could reorient the tetrahedra so that they all use the same scheme.  I was not very happy with either approach.